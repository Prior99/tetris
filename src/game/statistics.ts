import { Interval } from "utils";

export class Statistics {
    public pastIntervals: Interval[] = [];
    private nextInterval = new Interval(0);
    private currentTime = 0;
    public overallInterval = new Interval(0);

    public tick(seconds: number) {
        this.nextInterval.addTime(seconds - this.currentTime);
        this.overallInterval.addTime(seconds - this.currentTime);
        this.currentTime = seconds;
        if (this.nextInterval.completed) {
            this.pastIntervals.push(this.nextInterval);
            this.nextInterval = new Interval(this.currentTime);
        }
    }

    public reportScore(score: number) {
        this.nextInterval.score += score;
    }

    public reportLock() {
        this.nextInterval.reportLock();
        this.overallInterval.reportLock();
    }

    public get currentInterval(): Interval | undefined {
        return this.pastIntervals[this.pastIntervals.length - 1];
    }

    public reportLines(count: number) {
        this.nextInterval.reportLines(count);
        this.overallInterval.reportLines(count);
    }

    public lastN(n: number) {
        const intervals = [...this.pastIntervals, this.nextInterval];
        const result = Interval.combine(...intervals.slice(intervals.length - n, intervals.length));
        return result;
    }

    public get currentLocksPerMinute() {
        return this.overallInterval.locksPerMinute;
    }
}
