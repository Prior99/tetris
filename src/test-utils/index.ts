export * from "./mock-game";
export * from "./mock-playfield";
