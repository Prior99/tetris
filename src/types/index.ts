export * from "./atlas";
export * from "./constructable";
export * from "./cell-color";
export * from "./garbage";
export * from "./effects";
export * from "./page";
export * from "./game-mode";
export * from "./settings";
export * from "./sounds-mode";
export * from "./networking";
export * from "./mutable";
export * from "./game-parameters";
