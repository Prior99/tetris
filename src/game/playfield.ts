import { external, inject } from "tsdi";
import * as Random from "random-seed";
import { Config } from "config";
import { CellColor } from "types";
import { vec2, Matrix  } from "utils";

@external
export class Playfield extends Matrix {
    constructor(private random: Random.RandomSeed, @inject private config?: Config) {
        super(config!.logicalSize);
    }

    public addGarbageLines(lines: number) {
        this.update(this.moveUp(lines));
        for (let line = 0; line < lines; ++line) {
            const hole = this.random.range(this.config!.logicalSize.x);
            for (let x  = 0; x < this.dimensions.x; ++x) {
                if (x !== hole) {
                    this.put(vec2(x, line), CellColor.GARBAGE);
                }
            }
        }
    }

    public clear() {
        this.update(new Matrix(this.dimensions));
    }
}
