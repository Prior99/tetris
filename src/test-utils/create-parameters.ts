import { GameParameters, GarbageMode, GameMode, WinningConditionType } from "types";

export function createParameters(): GameParameters {
    return {
        seed: "some-seed",
        garbageMode: GarbageMode.HALF_REFERRED,
        gameMode: GameMode.SINGLE_PLAYER,
        initialGarbageLines: 2,
        initialLevel: 0,
        levelUpDisabled: false,
        winningCondition: { condition: WinningConditionType.HIGHEST_SCORE_ONE_GAME },
    };
}
