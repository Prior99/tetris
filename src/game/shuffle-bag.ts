import * as Random from "random-seed";
import {
    Tetrimino,
    TetriminoI,
    TetriminoJ,
    TetriminoL,
    TetriminoO,
    TetriminoS,
    TetriminoT,
    TetriminoZ,
} from "./tetriminos";
import { Playfield } from "./playfield";

export class ShuffleBag {
    private sequence: Tetrimino[] = [];
    private random: Random.RandomSeed;

    constructor(private playfield: Playfield, seed: string) {
        this.random = Random.create(seed);
        this.refill();
    }

    private shuffle<T>(array: T[]): T[] {
        const result: T[] = [];
        while (array.length > 0) {
            const [element] = array.splice(this.random.range(array.length), 1);
            result.push(element);
        }
        return result;
    }

    public take(): Tetrimino {
        if (!this.random) { throw new Error("Can't take from unitialized bag."); }
        const nextTetrimino = this.sequence.shift()!;
        if (this.sequence.length <= 7) { this.refill(); }
        nextTetrimino.refreshGhostPosition();
        return nextTetrimino;
    }

    public get preview(): Tetrimino[] {
        if (!this.random) { throw new Error("Can't peek into from unitialized bag."); }
        return this.sequence.slice(0, 6);
    }

    private refill() {
        this.sequence.push(...this.shuffle([
            new TetriminoI(this.playfield),
            new TetriminoJ(this.playfield),
            new TetriminoL(this.playfield),
            new TetriminoO(this.playfield),
            new TetriminoS(this.playfield),
            new TetriminoT(this.playfield),
            new TetriminoZ(this.playfield),
        ]));
    }
}
