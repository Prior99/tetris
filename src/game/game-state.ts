import { external, inject } from "tsdi";
import { bind } from "lodash-decorators";
import { Config } from "config";
import { vec2, Vec2, Matrix } from "utils";
import {
    AudioMoveDown,
    AudioRotateLeft,
    AudioRotateRight,
    AudioIncomingWarning,
    AudioIncomingCommitted,
    AudioHit,
    AudioScore1,
    AudioScore2,
    AudioScore3,
    AudioScore4,
    AudioLevelUp,
} from "resources";
import { Sounds } from "sounds";
import {
    Garbage,
    EffectType,
    GameParameters,
    GarbageMode,
    GameOverReason,
    WinningConditionType,
    CellColor,
} from "types";
import { ScoreAction, ScoreActionType, scorePointValue } from "./scoring";
import { Effects } from "./effects";
import { speed } from "./speed";
import { Tetrimino } from "./tetriminos";
import { Playfield } from "./playfield";
import { ShuffleBag } from "./shuffle-bag";

function calculateGarbage(lines: number): number {
    switch (lines) {
        case 1: return 0;
        case 2: return 1;
        case 3: return 2;
        default: return lines;
    }
}

@external
export class GameState {
    @inject private config: Config;
    @inject private sounds: Sounds;

    public current?: {
        tetrimino: Tetrimino;
        softDrops: number;
        hardDrops: number;
        usedHoldPiece: boolean;
    };
    public lastLockPosition?: Vec2;
    public lines = 0;
    public score = 0;
    public gameOverReason = GameOverReason.NONE;
    public holdPiece?: Tetrimino;
    public outgoingGarbage: Garbage[] = [];
    public incomingGarbage: Garbage[] = [];
    public comboCount = 0;
    public timeGameOver: undefined | number;

    private timeLastHit?: number;
    private timeLastLock?: number;
    private timeLastMoveDown = 0;
    private timeCurrent = 0;
    private toppedOutCount = 0;

    constructor(
        private shuffleBag: ShuffleBag,
        private playfield: Playfield,
        private effects: Effects,
        private parameters: GameParameters,
    ) {
        this.newTetrimino();
    }

    public tick(time: number) {
        this.timeCurrent = time;
        this.tickMatrix();
        this.tickGarbage();
        const timeOver = this.parameters.winningCondition.condition === WinningConditionType.SUM_IN_TIME &&
            time > this.parameters.winningCondition.seconds;
        if (timeOver) {
            this.setGameOver(GameOverReason.TIME_OVER);
        }
    }

    public get gameOver(): boolean {
        return this.gameOverReason !== GameOverReason.NONE;
    }

    private tickMatrix() {
        if (!this.current) { throw new Error("Ticked matrix but game was not initialized."); }
        if (this.gameOver) { return; }
        if (this.timeCurrent - this.timeLastMoveDown > this.speed || this.current.hardDrops) {
            this.timeLastMoveDown = this.timeCurrent;
            if (!this.current.tetrimino.hasHitFloor) {
                this.moveTetrimino();
            }
            if (this.current.tetrimino.hasHitFloor) {
                this.tickHitTetrimino();
            } else {
                this.timeLastHit = undefined;
                this.timeLastLock = undefined;
            }
        }
    }

    private tickHitTetrimino() {
        if (!this.hasHit) {
            this.timeLastHit = this.timeCurrent;
        }
        if (this.timeSinceLastHit! > this.config.lockTime) {
            this.commitTetrimino();
        }
    }

    private tickGarbage() {
        if (!this.current) { throw new Error("Ticked matrix but game was not initialized."); }
        const relevantGarbage = this.incomingGarbage.filter(garbage => this.isGarbageTimeout(garbage));
        if (relevantGarbage.length === 0) { return; }
        this.sounds.play(AudioIncomingCommitted);
        relevantGarbage.forEach(garbage => this.playfield.addGarbageLines(garbage.lines));
        this.incomingGarbage = this.incomingGarbage.filter(garbage => !this.isGarbageTimeout(garbage));
        this.current.tetrimino.refreshGhostPosition();
        this.current.tetrimino.moveSafeUp();
    }

    private isGarbageTimeout(garbage: Garbage): boolean {
        return this.timeCurrent - garbage.time > this.config.garbageTimeout;
    }

    public addIncomingGarbage(garbage: Garbage) {
        garbage.time = this.timeCurrent;
        this.incomingGarbage.push(garbage);
        this.sounds.play(AudioIncomingWarning);
    }

    public get speed() { return speed(this.level); }

    public get level() {
        if (this.parameters.levelUpDisabled) {
            return this.parameters.initialLevel;
        }
        return this.parameters.initialLevel + Math.floor(this.lines / 10);
    }

    @bind public inputRotateRight() {
        if (!this.current) { throw new Error("Received input event on uninitialized game state."); }
        this.sounds.play(AudioRotateRight);
        this.current.tetrimino.rotateRight();
    }

    @bind public inputRotateLeft() {
        if (!this.current) { throw new Error("Received input event on uninitialized game state."); }
        this.sounds.play(AudioRotateLeft);
        this.current.tetrimino.rotateLeft();
    }

    @bind public inputMoveLeft() {
        if (!this.current) { throw new Error("Received input event on uninitialized game state."); }
        this.current.tetrimino.moveLeft();
        this.sounds.play(AudioMoveDown);
    }

    @bind public inputMoveRight() {
        if (!this.current) { throw new Error("Received input event on uninitialized game state."); }
        this.current.tetrimino.moveRight();
        this.sounds.play(AudioMoveDown);
    }

    @bind public inputMoveDown() {
        if (!this.current) { throw new Error("Received input event on uninitialized game state."); }
        this.current.softDrops++;
        this.current.tetrimino.moveDown();
        this.sounds.play(AudioMoveDown);
    }

    @bind public inputHardDrop() {
        if (!this.current) { throw new Error("Received input event on uninitialized game state."); }
        this.current.hardDrops = this.current.tetrimino.hardDrop();
        this.commitTetrimino();
    }

    @bind public inputHoldPiece() {
        if (!this.current) { throw new Error("Received input event on uninitialized game state."); }
        if (this.current.usedHoldPiece) { return; }
        const currentHoldPiece = this.holdPiece;
        this.holdPiece = this.current.tetrimino;
        if (currentHoldPiece) {
            currentHoldPiece.reset();
            this.current = {
                tetrimino: currentHoldPiece,
                softDrops: 0,
                hardDrops: 0,
                usedHoldPiece: true,
            };
        } else {
            this.newTetrimino();
        }
    }

    public gameOverOtherUserHasWon() {
        this.setGameOver(GameOverReason.OTHER_USER_HAS_WON);
    }

    public gameOverLastManStanding() {
        this.setGameOver(GameOverReason.LAST_MAN_STANDING);
    }

    private moveTetrimino() {
        if (!this.current) { throw new Error("Tried to move tetrimino down on uninitialized game state."); }
        this.current.tetrimino.moveDown();
    }

    private reset() {
        this.lines = 0;
        this.comboCount = 0;
        this.timeLastHit = undefined;
        this.timeLastLock = undefined;
        this.holdPiece = undefined;
        this.lastLockPosition = undefined;
        this.playfield.clear();
    }

    private handleToppedOut() {
        this.toppedOutCount++;
        switch (this.parameters.winningCondition.condition) {
            case WinningConditionType.SUM_IN_TIME: {
                this.reset();
                break;
            }
            case WinningConditionType.BATTLE_ROYALE: {
                if (this.toppedOutCount >= this.parameters.winningCondition.lives) {
                    this.setGameOver(GameOverReason.TOPPED_OUT);
                } else {
                    this.reset();
                }
                break;
            }
            case WinningConditionType.CLEAR_GARBAGE:
            case WinningConditionType.HIGHEST_SCORE_ONE_GAME: {
                this.setGameOver(GameOverReason.TOPPED_OUT);
                break;
            }
        }
    }

    private newTetrimino() {
        const tetrimino = this.shuffleBag.take();
        if (tetrimino.isStuck) {
            this.handleToppedOut();
            tetrimino.refreshGhostPosition();
        }
        this.current = {
            tetrimino,
            softDrops: 0,
            hardDrops: 0,
            usedHoldPiece: false,
        };
    }

    private handleLevelUp() {
        this.sounds.play(AudioLevelUp);
    }

    private playScoreSound(count: number) {
        switch (count) {
            default:
            case 0: break;
            case 1: this.sounds.play(AudioScore1); break;
            case 2: this.sounds.play(AudioScore2); break;
            case 3: this.sounds.play(AudioScore3); break;
            case 4: this.sounds.play(AudioScore4); break;
        }
    }

    private cancelIncomingGarbage(count: number): number {
        let cancelled = 0;
        while (count > 0) {
            if (this.incomingGarbage.length === 0) { return cancelled; }
            const incoming = this.incomingGarbage[this.incomingGarbage.length - 1];
            cancelled += incoming.lines;
            if (incoming.lines > count) {
                incoming.lines -= count;
                break;
            } else {
                this.incomingGarbage.pop();
                count -= incoming.lines;
            }
        }
        return cancelled;
    }

    private calculateReturnedGarbage(clearedGarbageLines: number): number {
        switch (this.parameters.garbageMode) {
            case GarbageMode.FULL_REFERRED: return clearedGarbageLines;
            case GarbageMode.HALF_REFERRED: return Math.floor(clearedGarbageLines / 2);
            case GarbageMode.INITIAL_ONLY: return 0;
        }
        return 0;
    }

    private createOutgoingGarbage(count: number) {
        const clearedGarbageLines = this.cancelIncomingGarbage(count);
        const lines = calculateGarbage(count) + this.calculateReturnedGarbage(clearedGarbageLines);
        if (lines > 0 && this.parameters.garbageMode !== GarbageMode.NONE) {
            this.outgoingGarbage.push({ time: this.timeCurrent, lines });
        }
        if (this.comboCount >= 2) {
            this.effects.reportCombo(this.comboCount);
        }
    }

    private endCombo() {
        const { level, comboCount } = this;
        this.effects.clearCombo();
        this.comboCount = 0;
        if (this.comboCount >= 2) {
            this.awardScore({
                action: ScoreActionType.COMBO,
                level,
                comboCount,
            });
        }
    }

    private checkClearGarbageGameOver() {
        if (this.parameters.winningCondition.condition !== WinningConditionType.CLEAR_GARBAGE) {
            return;
        }
        if (!this.playfield.hasAny(CellColor.GARBAGE)) {
            this.setGameOver(GameOverReason.GARBAGE_CLEARED);
        }
    }

    private setGameOver(reason: GameOverReason) {
        this.gameOverReason = reason;
        this.timeGameOver = this.timeCurrent;
    }

    private commitTetrimino() {
        if (!this.current) { throw new Error("Can't commit tetrimino on uninitialized game state."); }
        this.sounds.play(AudioHit);
        const { tetrimino } = this.current;
        this.playfield.update(tetrimino.overlayedOnMatrix);
        const { matrix, offsets } = this.playfield.removeHorizontals();
        const count = offsets.length;
        if (count > 0) {
            this.comboCount++;
            this.playfield.update(matrix);
            const oldLevel = this.level;
            this.lines += count;
            if (this.level > oldLevel) {
                this.handleLevelUp();
            } else {
                this.playScoreSound(count);
            }
            this.scoreLineCount(count);
            this.createOutgoingGarbage(count);
            this.checkClearGarbageGameOver();
            offsets.forEach(y => this.effects.report({ effect: EffectType.LINE_CLEARED, y }));
        } else {
            this.endCombo();
        }
        this.lastLockPosition = tetrimino.offset.add(vec2(tetrimino.matrix.dimensions.x / 2, 0));
        this.newTetrimino();
        this.timeLastLock = this.timeCurrent;
        this.timeLastHit = undefined;
    }

    public get timeSinceLastHit(): number | undefined {
        if (!this.timeLastHit) { return; }
        return this.timeCurrent - this.timeLastHit;
    }

    public get timeSinceLastLock(): number | undefined {
        if (!this.timeLastLock) { return; }
        return this.timeCurrent - this.timeLastLock;
    }

    public get hasHit(): boolean {
        return this.timeLastHit !== undefined;
    }

    public awardScore(action: ScoreAction) {
        this.score += scorePointValue(action);
    }

    public scoreLineCount(count: number) {
        if (!this.current) { throw new Error("Can't score on uninitialized game state."); }
        this.effects.reportLines(count);
        const { level } = this;
        switch (count) {
            case 1: this.awardScore({ action: ScoreActionType.SINGLE, level });
            case 2: this.awardScore({ action: ScoreActionType.DOUBLE, level });
            case 3: this.awardScore({ action: ScoreActionType.TRIPLE, level });
            case 4: this.awardScore({ action: ScoreActionType.TETRIS, level });
        }
        if (this.current.hardDrops) {
            this.awardScore({ action: ScoreActionType.HARD_DROP, cells: this.current.hardDrops });
        }
        if (this.current.softDrops) {
            this.awardScore({ action: ScoreActionType.SOFT_DROP, cells: this.current.softDrops });
        }
    }

    public get temporaryState(): Matrix {
        if (!this.current) {
            return this.playfield;
        }
        return this.current.tetrimino.overlayedOnMatrixWithGhost;
    }
}
