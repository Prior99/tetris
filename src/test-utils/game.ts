import { merge } from "ramda";
import { GameMode, Mutable, WinningConditionType, GarbageMode } from "types";
import { vec2 } from "utils";
import { Game } from "game";
import { mockPlayfield } from "./playfield";

export function mockGame(override?: Partial<Game>): Mutable<Game> {
    const defaults: Game = {
        running: true,
        serial: "random string",
        seconds: 10,
        tetriminoPreviews: [],
        level: 1,
        lines: 0,
        score: 0,
        gameOver: false,
        tetriminoOffset: vec2(4, 4),
        lastLockPosition: vec2(5, 5),
        hasHit: false,
        timeSinceLastLock: undefined,
        timeSinceLastHit: undefined,
        timeSinceLastDouble: 0.3,
        timeSinceLastTriple: undefined,
        timeSinceLastTetris: undefined,
        timeSinceComboStart: undefined,
        timeSinceComboEnd: undefined,
        timeout: undefined,
        comboCounts: [],
        temporaryState: mockPlayfield(),
        effects: [],
        restart: jest.fn(),
        addIncomingGarbage: jest.fn(),
        hasOutgoingGarbage: false,
        outgoingGarbage: [],
        clearOutgoingGarbage: jest.fn(),
        start: jest.fn(),
        stop: jest.fn(),
        tick: jest.fn(),
        incomingGarbage: [],
        holdPiece: undefined,
        latestIncomingGarbage: undefined,
        parameters:{
            winningCondition: {
                condition: WinningConditionType.CLEAR_GARBAGE,
            },
            gameMode: GameMode.SINGLE_PLAYER,
            seed: "some-seed",
            initialGarbageLines: 10,
            garbageMode: GarbageMode.NONE,
            initialLevel: 0,
            levelUpDisabled: false,
        }
    } as any;
    return merge(defaults, override ?? {}) as any as Mutable<Game>;
}
