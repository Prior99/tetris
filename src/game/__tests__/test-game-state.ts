import { GameParameters, GameOverReason, WinningConditionType } from "types";
import { createParameters } from "test-utils";
import { GameState } from "../game-state";
import { ShuffleBag } from "../shuffle-bag";
import { Playfield } from "../playfield";
import { Effects } from "../effects";
import { Statistics } from "../statistics";

describe("GameState", () => {
    let shuffleBag: ShuffleBag;
    let playfield: Playfield;
    let effects: Effects;
    let parameters: GameParameters;
    let gameState: GameState;

    beforeEach(() => {
        parameters = createParameters();
        parameters.winningCondition = { condition: WinningConditionType.CLEAR_GARBAGE };
        playfield = new Playfield(parameters);
        shuffleBag = new ShuffleBag(playfield, parameters.seed);
        effects = new Effects();
        gameState = new GameState(shuffleBag, playfield, effects, parameters, new Statistics());
    });

    describe("initially", () => {
        it("shows the expected temporary state", () => expect(gameState.temporaryState.toString()).toMatchSnapshot());
    });

    describe("after 2s", () => {
        beforeEach(() => gameState.tick(2));

        it("shows the expected temporary state", () => expect(gameState.temporaryState.toString()).toMatchSnapshot());
    });

    describe("after topping out", () => {
        beforeEach(() => {
            for (let i = 0; i < 9; ++i) { gameState.inputHardDrop(); }
        });

        it("shows the expected temporary state", () => expect(gameState.temporaryState.toString()).toMatchSnapshot());

        it("sets game over reason", () => expect(gameState.gameOverReason).toBe(GameOverReason.TOPPED_OUT));
    });

    describe("dropping first tetrimino", () => {
        beforeEach(() => {
            gameState.inputMoveLeft();
            gameState.inputMoveLeft();
            gameState.inputHardDrop();
        });

        it("shows the expected temporary state", () => expect(gameState.temporaryState.toString()).toMatchSnapshot());

        describe("holding second tetrimino", () => {
            beforeEach(() => gameState.inputHoldPiece());

            it("shows the expected temporary state", () => {
                expect(gameState.temporaryState.toString()).toMatchSnapshot();
            });

            describe("rotating and dropping third tetrimino", () => {
                beforeEach(() => {
                    gameState.inputRotateRight();
                    gameState.inputRotateRight();
                    gameState.inputMoveRight();
                    gameState.inputHardDrop();
                });

                it("shows the expected temporary state", () => {
                    expect(gameState.temporaryState.toString()).toMatchSnapshot();
                });

                it("reports the cleared line", () => expect(gameState.lines).toBe(1));

                it("reports the score", () => expect(gameState.score).toBe(36));

                describe("dropping fourth tetrimino", () => {
                    beforeEach(() => gameState.inputHardDrop());

                    it("shows the expected temporary state", () => {
                        expect(gameState.temporaryState.toString()).toMatchSnapshot();
                    });

                    it("reports the cleared line", () => expect(gameState.lines).toBe(2));

                    it("reports the score", () => expect(gameState.score).toBe(76));

                    it("reports the correct game over reason", () => {
                        expect(gameState.gameOverReason).toBe(GameOverReason.GARBAGE_CLEARED);
                    });

                    it("sets combo count to two", () => expect(gameState.comboCount).toBe(2));
                });
            });
        });
    });
});
