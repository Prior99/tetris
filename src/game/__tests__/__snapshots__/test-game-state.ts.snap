// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameState after 2s shows the expected temporary state 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 4 4 0 0 0
0 0 0 0 0 4 4 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 8 8 0 0 0
0 0 0 0 0 8 8 0 0 0
9 9 9 9 9 9 9 9 0 9
9 9 9 9 9 0 9 9 9 9
"
`;

exports[`GameState after topping out shows the expected temporary state 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 1 7 0 0 0
0 0 0 0 0 1 7 7 0 0
0 0 0 0 0 1 5 5 0 0
0 0 0 0 0 5 5 3 0 0
0 0 0 0 0 3 3 3 0 0
0 0 0 0 0 0 6 0 0 0
0 0 0 0 0 6 6 6 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 2 0 0 0 0
0 0 0 0 0 2 2 2 0 0
0 0 0 0 0 7 7 0 0 0
0 0 0 0 0 0 7 7 0 0
0 0 0 0 0 4 4 0 0 0
0 0 0 0 0 4 4 0 0 0
9 9 9 9 9 9 9 9 0 9
9 9 9 9 9 0 9 9 9 9
"
`;

exports[`GameState dropping first tetrimino holding second tetrimino rotating and dropping third tetrimino dropping fourth tetrimino shows the expected temporary state 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 6 0 0 0
0 0 0 0 0 6 6 6 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 8 0 0 0
0 0 0 0 0 8 8 8 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 4 4 1 0 0 0 0
0 0 0 4 4 1 2 2 2 0
"
`;

exports[`GameState dropping first tetrimino holding second tetrimino rotating and dropping third tetrimino shows the expected temporary state 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 1 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 8 0 0 0 0
0 0 0 4 4 8 0 0 0 0
0 0 0 4 4 8 2 2 2 0
9 9 9 9 9 8 9 9 9 9
"
`;

exports[`GameState dropping first tetrimino holding second tetrimino shows the expected temporary state 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 2 0 0 0 0
0 0 0 0 0 2 2 2 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 4 4 8 0 0 0 0
0 0 0 4 4 8 8 8 0 0
9 9 9 9 9 9 9 9 0 9
9 9 9 9 9 0 9 9 9 9
"
`;

exports[`GameState dropping first tetrimino shows the expected temporary state 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 7 7 0 0 0
0 0 0 0 0 0 7 7 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 4 4 8 8 0 0 0
0 0 0 4 4 0 8 8 0 0
9 9 9 9 9 9 9 9 0 9
9 9 9 9 9 0 9 9 9 9
"
`;

exports[`GameState initially shows the expected temporary state 1`] = `
"0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 4 4 0 0 0
0 0 0 0 0 4 4 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 0 0 0 0 0
0 0 0 0 0 8 8 0 0 0
0 0 0 0 0 8 8 0 0 0
9 9 9 9 9 9 9 9 0 9
9 9 9 9 9 0 9 9 9 9
"
`;
